tensorflow/python/ops/image_ops_impl.py

```diff
@@ -1483,13 +1483,13 @@ def is_jpeg(contents, name=None):
 
 
 @tf_export('image.decode_image')
-def decode_image(contents, channels=None, name=None):
+def decode_image(contents, channels=None, dtype=dtypes.uint8, name=None):
   """Convenience function for `decode_bmp`, `decode_gif`, `decode_jpeg`,
   and `decode_png`.
 
   Detects whether an image is a BMP, GIF, JPEG, or PNG, and performs the
   appropriate operation to convert the input bytes `string` into a `Tensor` of
-  type `uint8`.
+  type `uint8`. PNG also supports type `uint16`.
```

Comment:
(From API review)

So, IIUC, the contract of this function is that `dtype` will be ignored if the image doesn't happen to be a PNG?

It seems a more appropriate contract would be to always respect the `dtype` argument (perhaps by casting to `tf.uint16` if necessary for the non-PNG case)

Reply:
The tf.image.decode_image function is based on tf.image.decode_*,
only tf.image.decode_png have `dtype` argument.

* : bmp, gif, jpeg, or png.

Reply:
I understand, but here we're talking about the API for `tf.image.decode_image`.

With the change as proposed:
`tf.decode_image(image_file, dtype=tf.uint16)` will return a tensor with an unpredictable `dtype`, which is not a great API.

The suggestion is to make it so that the output of `tf.decode_image` *always* respects the `dtype` argument provided.

As suggested above, this can be achieved by calling `tf.cast` on the output of  `tf.image.decode_{bmp,gif,jpeg}`, since they all return a `tf.uint8` tensor.

Sound reasonable?